File: Python/TwoSum.py
class Solution(object):
    def twoSum(self, nums, target):
        """
        :type nums: List[int]
        :type target: int
        :rtype: List[int]
        """
        map={}
        l=len(nums)
        for i in range(l):
            map[nums[i]]=i
        for i in range(l):
            j=target-nums[i]
            if map.has_key(j):
                if i!=map[j]:
                    if map[j]>i:
                        return (i+1,map[j]+1)
                    else:
                        return (map[j]+1,i+1)
s=Solution()
print s.twoSum([2,7,11,15],9)

File: Java/src/algorithms/TwoSum.java
package algorithms;

import java.util.HashMap;

/**
 * @author yuchengkai
 *
 * @time   2015年12月19日
 *
 */
public class TwoSum {
	
	/**
	 * Use a map to record the numbers and their indices.
	 * Iterate the array again to check if there is target-i in the map for 
	 * every i in the array. 
	 * */
    public int[] twoSum(int[] nums, int target) {
        HashMap<Integer,Integer> m=new HashMap<Integer,Integer>();
        int[] r=new int[2];
        int l=nums.length;
        for(int i=0;i<l;i++){
        	m.put(nums[i], i);
        }
        for(int i=0;i<l;i++){
        	if(m.containsKey(target-i)){
        		int j=m.get(target-i);
        		if(i>j){
        			r[0]=j;
        			r[1]=i;
        		}else{
        			r[0]=i;
        			r[1]=j;
        		}
        		break;
        	}
        }
        return r;
    }
}
